src/odemis/driver/smaract.py

```diff
@@ -3084,3 +3085,1139 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+        # since even small changes in the experimental setup or environment might cause the old
+        # adjustment parameters to become non-ideal. Therefore, the values for all channels in .referenced
+        # will be False, until the full referencing routine has been performed.
+        # These procedure can take a while (up to 10 minutes), especially after the power of the system
+        # has just been turned on. Therefore, we don't wait until the procedure is complete.
+        # Starting from standby is generally much faster.
+        self._executor = CancellableThreadPoolExecutor(1)  # one task at a time
+        channel_ref = {ch: False for ch in self.channels}
+        self.referenced = model.VigilantAttribute(channel_ref, readonly=True)  # VA dict str(channel) -> bool
+        if ref_on_init:
+            f = self.reference()
+        else:
+            f = self._validate()
+            logging.warning("Picoscale is not referenced. The reported positions might not be accurate "
+                            "until referencing occurs.")
+        f.add_done_callback(self._on_referenced)
+
+    def terminate(self):
+        # should be safe to close the device multiple times if terminate is called more than once.
+        self._polling_thread.cancel()
+        if self._executor:
+            self.stop()
+            self.core.SA_SI_Close(self._id)
+            self._executor = None
+        super(Picoscale, self).terminate()
+
+    @staticmethod
+    def scan():
+        """
+        Find all available Picoscale controllers.
+        returns: (set of str) locator strings
+        """
+        core = SA_SIDLL()
+        b_len = 1024
+        buf = create_string_buffer(b_len)
+        core.SA_SI_EnumerateDevices(c_char_p(b""), buf, byref(c_size_t(b_len)))
+        locators = buf.value.decode('latin1').split("\n")
+        return set(locators)  # sometimes the same locator is returned multiple times, convert to set for unique values
+
+    @staticmethod
+    def _openConnection(locator):
+        """
+        Open usb/ethernet connection to device.
+        locator: (str) Use "fake" for a simulator.
+            For a real device, Picoscale controllers with USB interface can be addressed with the
+            following locator syntax:
+                usb:ix:<id>
+            where <id> is the first part of a USB devices serial number which
+            is printed on the Picoscale controller.
+            If the controller has a TCP/IP connection, use:
+                network:<ip>:<port>
+            The device can also be addressed by its serial number:
+                usb:sn:<serial_number>
+                network:sn:<serial_number>
+        returns: (int) device handle for API functions
+        """
+        if locator == "fake":
+            return 0
+
+        # TODO: try multiple times, first time sometimes doesn't work
+        logging.debug("Connecting to locator %s.", locator)
+        core = SA_SIDLL()
```

Comment:
Still, I'd rather try to reuse the same core as in the init, firstly to make sure the simulator is also using this code, and secondly to avoid opening twice the DLL at init.

Reply:
How about this: I could move the "core" part into the `_openConnection` function and return it with the id:
``` self.core, self._id = self._openConnection(locator)```
And in `_openConnection(locator)`:
```
        if locator == "fake":
            core = FakePicoscale_DLL()
            return core, 0
        else:
            core = SA_SIDLL()
         ...
         return core, id
```